[CFEP-13] secure package uploads (#24)
conda-forge/cfep: cfep-13.md

<table>
<tr><td> Title </td><td> Securing `conda-forge` Uploads to `anaconda.org` </td>
<tr><td> Status </td><td> Proposed </td></tr>
<tr><td> Author(s) </td><td> Matt, Marius, Anthony, Isuru</td></tr>
<tr><td> Created </td><td> Feb 27, 2020</td></tr>
<tr><td> Updated </td><td> Feb 27, 2020</td></tr>
<tr><td> Discussion </td><td> NA </td></tr>
<tr><td> Implementation </td><td> NA </td></tr>
</table>

## Abstract

Currently the `anaconda.org` tokens on each feedstock allow any `conda-forge`
feedstock maintainer to upload any output to any `anaconda.org` package.
This could allow someone to push a compromised and/or malicious build
of say the `python` package. This situation is a security risk that we
should mitigate.

## Specification

We will secure the uploads of outputs to `anaconda.org` via the following process.

1. A public, global registry of all outputs from all feedstocks will be kept
   in a repo on github. Write access to this repo is limited to core and bot 

>    admins as well.

Reviewer: Could you please clarify what is being uploaded here? Packages? Something else?

Author: this Is an implementation detail and doesn’t belong in the cfep. FWIW I have used json blobs for now.

Author: Literally a record of the output names each feedstock generates.

Author: Hows that?

Reviewer: Thanks that was what I was looking for 🙂

2. When a feedstock builds new outputs it wants to upload, it will push these outputs
   to a staging organization on `anaconda.org`. These outputs will either be signed 
   so that they can be validated or the feedstock will directly alert the validation 
   service in a secure, verifiable manner.
3. The validation service will validate that any outputs in the staging org are 

>    from the correct feedstock, the actual outputs to be upoloaded, and allowed 

Reviewer: ```suggestion
   from the correct feedstock, the actual outputs to be uploaded, and allowed 
```

> 3. When a feedstock builds new outputs it wants to upload, it will push these outputs

Reviewer 2: ```suggestion
4. When a feedstock builds new outputs it wants to upload, it will push these outputs
```

   to a staging organization on `anaconda.org`. It will also send the names of the
   outputs, their SHA256 checksums, the feedstock name, and the secret key to a
   validation web service. This web service then verifies the secret key-feedstock
   combination and verifies that the outputs are allowed for that feedstock.
4. If everything is ok, the outputs are copied from the staging organization to the
   main conda-forge channels.

> 6. The service should post a comment on the PR that the outputs were copied.

Reviewer 2: How about just waiting for the webservice to report that the package was copied?

Author: Not sure what you mean. Once the web service has copied, it should report back to let people know.

Reviewer 2: In CI, we send a request to the webservice and the webservice can send a 200 code if the package was copied and a 403 if it's forbidden. CI could fail if the code is not 200.

Author: Ah I see. That is fine for the CI, but it doesn't really help people understand why something failed. I'd like things to be transparent enough so that users know which thing to go track down when we inevitably don't copy something due to a bug, etc.

### The Impact on Users

In order to mitigate the impact on feedstock maintainers, we will

1. Advertise this change widely as it is implemented.

> 2. Create an admin bot command to create a PR for new outputs.

Reviewer 2: Do we need this? Maybe the webservice should just create a PR for new outputs.

Author: Well in principle, the outputs need to be in the database before the webservice attempts to copy them. So the idea here is that if a user gets a linter warning that an output is not registered, they can use the admin command to generate the PR to do so. This PR will get merged automatically if nothing conflicts.

> 3. Add a linter check that all outputs exist.

Reviewer 2: What does this mean?

Author: Well for a package to be uploaded, all of its outputs must be in the global database. The linter should check this and warn users if not.

Reviewer 2: ```suggestion
3. Add a linter check that all outputs are registered for this feedstock in the global database.
```

> 3. Add a linter check that all outputs are registered for this feedstock in the global database.

Reviewer 2: This would make the linter as slow at the rerender command because the package has to be rendered for all the platforms.

Author: Hmmmm. We could allow the CI on PRs to do a test check of the outputs against the web service and the CI can fail for PRs if outputs are not registered.

Author: My main thing is that something somewhere needs to report in a visible way to users what they need to do.

Author: We could have the rerender command do this check and post a comment. That is slow anyways and actually renders the recipe.

Reviewer 2: +1. I would remove the linter check then add something about using CI for this.

Reviewer 2: > We could have the rerender command do this check and post a comment. That is slow anyways and actually renders the recipe.

You mean run rerender on every push to a PR? That would cripple the webservice

Author: No I mean when it happens to be called. So if a user happens to use the rerender service to rerender a feedstock, we go ahead and check the outputs then.

Author: The CI should check too.

Reviewer 2: That's too much complexity. Just CI is enough IMO.

Author: fair enough - I guess it can post a loud error message that people can look at

3. A public, global registry of all outputs from all feedstocks will be kept
   in a repo on github. Write access to this repo is limited to core as well.
4. When a feedstock builds new outputs it wants to upload, it will push these outputs
   to a staging organization on `anaconda.org`. It will also send the names of the
   outputs, their SHA256 checksums, the feedstock name, and the secret key to a
   validation web service. This web service then verifies the secret key-feedstock
   combination and verifies that the outputs are allowed for that feedstock.
5. If everything is ok, the outputs are copied from the staging organization to the
   main conda-forge channels.
6. The service should fail the CI run if the output is not copied. It should 
   also post a comment on the PR that the outputs were copied or not to help users.

### The Impact on Users

In order to mitigate the impact on feedstock maintainers, we will

1. Advertise this change widely as it is implemented.
2. Create an admin bot command to create a PR for new registering new outputs.
3. PR CI runs should check all of their outputs against the web service in a "dry-run"

>    mode and fail if any output is not registered.

Reviewer 2: If there's no overlap, we shouldn't fail

Author: Well something somewhere has to push to the registry of outputs and this has to be gated in some way.

Author: The failure is here to indicate to a user that they need to do something to register their outputs.

Reviewer 2: The webservice that validates the outputs can push to the registry if there's no overlap. This way, only users who add outputs that conflict needs to do anything. Others can be oblivious of this change.

Author: I guess given everything else the dry run mode could simply register outputs that don't exist and only fail if there is a conflict.

Author: this is no different than automerge

Author: ditto for uploads

Reviewer 2: > this is no different than automerge

Exactly. I wouldn't register in a PR though as people make mistakes.

Author: Ack sorry I missed your comments @isuruf! Yup I agree!

> 3. Both the "dry run" and "upload" commands to the web service will automatically 

Reviewer 2: I'm not a fan of registering in `dry run` mode as people make mistakes in the PRs or change their minds about naming.

Author: good point

>    register new outputs for a feedstock that do not conflict with any other outputs.

Reviewer 2: ```suggestion
   register new outputs for a feedstock that do not conflict with any other feedstock.
```

Author: Well it is the outputs of any other feedstock.

Author: take a look at the new text which combines both things and is most clear I hope

>    don't overlap with any existing name on any feedstock.

Reviewer 2: Do we need PRs in this case? Just push to master?

Author: Yes we can have the admin command push straight to master if there is no overlap.

<table>
<tr><td> Title </td><td> Securing `conda-forge` Uploads to `anaconda.org` </td>
<tr><td> Status </td><td> Proposed </td></tr>
<tr><td> Author(s) </td><td> Matt, Marius, Anthony, Isuru</td></tr>
<tr><td> Created </td><td> Feb 27, 2020</td></tr>
<tr><td> Updated </td><td> Feb 27, 2020</td></tr>
<tr><td> Discussion </td><td> NA </td></tr>
<tr><td> Implementation </td><td> NA </td></tr>
</table>

## Abstract

Currently the `anaconda.org` tokens on each feedstock allow any `conda-forge`
feedstock maintainer to upload any output to any `anaconda.org` package.
This could allow someone to push a compromised and/or malicious build
of say the `python` package. This situation is a security risk that we
should mitigate.

## Specification

We will secure the uploads of outputs to `anaconda.org` via the following process.

1. Each feedstock is provisioned with a secret key that is only known to the
   maintainers of that feedstock and the core team. These keys will be kept in

>    a private repo with only core members having write access. This key will be pushed to

Reviewer 2: I don't think we need maintainers any access at all. They'd have indirect access via CI, but that's it.

```suggestion
   a private repo with only core members having read and write access. This key will be pushed to
```

Author: Yes this was my intent with the phrase "private repo" but it was not clear. 100% agreed!

   the CI providers for that repo as well.

> 2. Feedstock creation will move to a separate process in order to secure the keys

Reviewer 2: What does this mean? Why can't we have the same process as now?

Author: Well @mariusvniekerk seemed to suggest on gitter that this would be a good change, but I agree the exact technical issue here is not clear to me. 

We do need to generate and register new keys to the private repo in a secure way. IDK if having this done with code on staged recipes is a great idea. 

However, we don't have to redo staged recipes to address this. We could have the web service watch repo creation events in the conda-forge org and automatically register a key when it detects a new feedstock.

Author: We also need to add the keys to feedstock CI services in a secure way. Again it is not 100% clear to me how we do this. Maybe the webservice has a register command we call.

Reviewer 2: Why can't we have the feedstock creation script in `staged-recipes` do this?

Author: The actual constraint is that we have to avoid leaking the secret key we generate to the outside world. 

The question of who can see those keys is separate at some level. Having them be generated in staged recipes would allow core+any staged recipe maintainer be able to see the keys as they are made in principle.

If we are ok with that, then staged-recipes is fine. 

After thinking about it for a while and remembering that I was not always on core, I think I'd still prefer to not have even our encrypted tokens for any service available outside of core. However, I agree this is a different issue that is orthogonal to this CFEP.

Reviewer 2: I agree that this is orthogonal to this CFEP. (If we go this route, we'd need to remove the feedstock creation scripts from staged-recipes as they contain github tokens, CI tokens, etc.)

>    the secret keys and push then to the CI providers.

Reviewer 3: Shouldn't this be in keybase rather than on github?

Author: We can do this, but IDK how to push and pull from keybase with a key securely when that key is stored on staged-recipes.

Author: We may not want keys for keybase itself floating around in our CI scripts.

Reviewer 3: hmm yeah, I see

2. A public, global registry of all outputs from all feedstocks will be kept
   in a repo on github. Write access to this repo is limited to core and bot 
   admins as well.
3. When a feedstock builds new outputs it wants to upload, it will push these outputs
   to a staging organization on `anaconda.org`. It will also send the names of the

>    outputs, their SHA256 checksums, the feedstock name, and the secret key to a

Reviewer 4: Are we literally sending the secret key? Better to use public/private keys and sign the message to the web service, right? Just to avoid having the key itself traveling around as much as possible.

Author: A lot of services send API keys over https these days (in the authorization bearer headers), so I figured that would be fine here. Are you saying that is not secure enough for this purpose?

Reviewer 4: I mean, I suppose https is encrypted anyway, and I'm definitely no expert on these things. It just feels more potentially risky in case of accidentally sending requests to the wrong place / etc.

Author: Hmmm. At a practical level, either an API key or the private key (to sign messages) needs to be stored on the CI services. For that point of view the risk for spilling it is mostly the same and due to the CI service itself, maintainers making edits etc IIUIC. 

I am not personally familiar with how to store a private key on the CI services or if they are setup to handle such things. Should we put it in an env var?

Unless we have a good answer for how to store a private key, I’m inclined to go the API key route since I don’t think I will mess it up as easily.

Reviewer 5: From reading the text, it is not clear to me what the secret key is used for. Is this for signing? Then using async signing should allow us to have a public key for that?

Author: The secret key is a standard ApI key. 

Do you know how to put private keys in the CI services?

If we embed an entire file in an env var, will the CO service properly mask it in outputs?

Reviewer 5: You can store private keys in CI services, depends a bit on the service. Travis provides a different public encryption key per repo, for Github Actions you can set it via the UI in Actions Tab (and there is hopefully an API for that), not sure about Azure but I would expect to provide a similar thing. They all will mask the exact value which is in the key in the CI log. The key will also be provided as an env var in the CI job but only when run as part of a commit in `conda-forge`, not in PRs.

Author: This entire issue around an API key versus public/private keys is a red herring. 

My main concern is correctly implementing this system both in the code and as a whole with various keys secured correctly. 

I’m assuming I’ll be doing the coding here and I am not confident I can handle the public/private key option correctly. 

A worse outcome is for us to think we have fixed this when there is actually a bug. This is especially true with respect to cryptography. 

If someone else wants to help code this with public/private keys and knows definitely how to secure the private key on our CI services, then I am happy to do that route. 

Otherwise let’s use an API key sent over https since that can be done by me confidently.

Author: Setting keys via a UI is a nonstarter. 

It’s not the env var I’m worried about @xhochy. Do you know for sure that an env var with new lines will be masked properly in all of our CI services? Usually people stuff API keys in these, not full files.

Reviewer 5: > It’s not the env var I’m worried about @xhochy. Do you know for sure that an env var with new lines will be masked properly in all of our CI services? Usually people stuff API keys in these, not full files.

With full files, I have no experience, so: no.

Reviewer 6: If the worry is about multi-line text, it could probably be base64-encoded for env var storage, no?

Author: IDK what you are talking about, which is actually the point here. If you have experience doing this correctly and want to proactively help implement this CFEP, then, by all means, lets go for it.

Reviewer 6: Unfortunately I absolutely do not have the spare cycles to help in any concrete way :-( For completeness, [base64](https://en.wikipedia.org/wiki/Base64) encodings can help pack arbitrary binary data into things like environment variables.

Author: cool - let's go with the API key approach for now. If you get spare time and have interest, then feel free to pitch in!

Reviewer 7: `base64` might not be the best choice if the CI would decide to only mask the encoded but not the decoded key :laughing:.
How ever you want to go with it, this would also work:
```sh
# put it in one line:
key_as_one_line="$( < key-file head -n-1 | tail -n+2 | tr -d \\n )"
# reassemble and use:
echo -e "-----BEGIN PRIVATE KEY-----\n${key_as_one_line}\n-----END PRIVATE KEY-----" | \
  openssl dgst -sha256 -sign - some-file > some-file.sha256
```

   validation web service. This web service then verifies the secret key-feedstock
   combination and verifies that the outputs are allowed for that feedstock.
4. If everything is ok, the outputs are copied from the staging organization to the

>    main conda-forge channels.

Reviewer 3: We should probably delete bad/old artifacts from the staging channel after some period of time (28 days?).  I don't want the staging channel to be a usuable mirror of conda-forge, or else people could use that and it would be a security hole

Reviewer 2: After copying, we could delete right?

Author: Yeah we can do that. But this is on them. They should not use an unsecured channel. We cannot control what they do.

Reviewer 3: I think yeah we could delete after a copy

Author: done.
